### PR TITLE
implement order logic

### PR DESCRIPTION
**Pull request by don-bigdad**

### CODE REVIEW

File: src/main/java/com/example/springbootbookshop/controller/OrderController.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    @Operation(summary = "Get special order from user history")
+    @PreAuthorize("hasRole('USER')")
+    public Set<OrderItemDto> showOrder(Authentication authentication,
+                                       @PathVariable Long orderId) {
+        return orderService.getOrder(getUserId(authentication), orderId);
+    }
+
+    private Long getUserId(Authentication authentication) {
+        User user = (User) authentication.getPrincipal();
+        return user.getId();
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/springbootbookshop/controller/OrderController.java]: can be an inline variable

File: src/main/java/com/example/springbootbookshop/controller/OrderController.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    @PreAuthorize("hasRole('USER')")
+    public CartDto makeOrder(Authentication authentication,
+                             @RequestBody RequestOrderDto requestOrderDto) {
+        return orderService.save(getUserId(authentication), requestOrderDto);
+    }
+
+    @PatchMapping("/{orderId}")
+    @Operation(summary = "Update order status (for admin only)")
+    @PreAuthorize("hasRole('ADMIN')")
+    public void updateStatus(@Positive @PathVariable Long orderId,
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/springbootbookshop/controller/OrderController.java:47]: Positive doesn’t  work without Validated

File: src/main/java/com/example/springbootbookshop/controller/OrderController.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+                             @RequestBody UpdateRequestOrderStatus updateStatus) {
+        orderService.updateOrderStatus(orderId, updateStatus);
+    }
+
+    @GetMapping
+    @Operation(summary = "Get order history")
+    @PreAuthorize("hasRole('USER')")
+    public Set<OrderDto> showOrderHistory(Authentication authentication) {
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/springbootbookshop/controller/OrderController.java]:
```suggestion
    public Set<OrderDto> getOrderHistory(Authentication authentication) {
```

File: src/main/java/com/example/springbootbookshop/controller/OrderController.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+        return orderService.showAllOrders(getUserId(authentication));
+    }
+
+    @GetMapping("/{orderId}/items/{itemId}")
+    @Operation(summary = "Get special item from order history")
+    @PreAuthorize("hasRole('USER')")
+    public OrderItemDto showOrderItem(Authentication authentication,
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/springbootbookshop/controller/OrderController.java]: the same

File: src/main/java/com/example/springbootbookshop/entity/Order.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    private BigDecimal total;
+
+    @Column(nullable = false)
+    private LocalDateTime orderDate;
+
+    @Column(nullable = false)
+    private String shippingAddress;
+
+    @OneToMany(mappedBy = "order", orphanRemoval = true,
+            cascade = CascadeType.REMOVE)
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/springbootbookshop/entity/Order.java]: Maybe  ALL?

File: src/main/java/com/example/springbootbookshop/mapper/OrderItemMapper.java [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+import org.mapstruct.Mapper;
+import org.mapstruct.Mapping;
+import org.mapstruct.NullValueCheckStrategy;
+
+@Mapper(componentModel = "spring",
+        injectionStrategy = InjectionStrategy.CONSTRUCTOR,
+        nullValueCheckStrategy = NullValueCheckStrategy.ALWAYS,
+        implementationPackage = "<PACKAGE_NAME>.impl")
+public interface OrderItemMapper {
+
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/springbootbookshop/mapper/OrderItemMapper.java]:
```suggestion
```

File: src/main/java/com/example/springbootbookshop/repository/OrderItemsRepository.java [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+import com.example.springbootbookshop.entity.Order;
+import com.example.springbootbookshop.entity.OrderItem;
+import java.util.Optional;
+import java.util.Set;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.stereotype.Repository;
+
+@Repository
+public interface OrderItemsRepository extends JpaRepository<OrderItem, Long> {
+    Optional<OrderItemDto> findByOrderAndId(Order order, Long id);
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/springbootbookshop/repository/OrderItemsRepository.java]: Why do you return Dto from DB?

File: src/main/java/com/example/springbootbookshop/repository/OrderItemsRepository.java [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+
+    Set<OrderItemDto> findByOrder(Order order);
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/springbootbookshop/repository/OrderItemsRepository.java]: the same

File: src/main/java/com/example/springbootbookshop/service/impl/CartServiceImpl.java [hunk 9]
```diff
@@ -46,13 +46,14 @@ public CartItemDto addItemToCart(Long userId, RequestCartItemDto requestCartItem
...
-                });
-        cartItem.setQuantity(quantityToAdd + cartItem.getQuantity());
+                .orElse(null);
+        if (cartItem == null) {
+            cartItem = cartItemMapper.toEntity(requestCartItemDto);
+            cartItem.setBook(book);
+            cartItem.setCart(cart);
+        } else {
+            cartItem.setQuantity(cartItem.getQuantity() + quantityToAdd);
+        }
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/springbootbookshop/service/impl/CartServiceImpl.java]: Can you replace it using a ternary operator?

File: src/main/java/com/example/springbootbookshop/service/impl/OrderServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+                    orderItem.setBook(cartItem.getBook());
+                    orderItem.setPrice(cartItem.getBook().getPrice()
+                            .multiply(BigDecimal.valueOf(cartItem.getQuantity())));
+                    orderItem.setQuantity(cartItem.getQuantity());
+                    return orderItem;
+                })
+                .forEach(orderItemsRepository::save);
+
+        cartItemsRepository.deleteAll(cartItems);
+        return cartMapper.toDto(cart);
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/springbootbookshop/service/impl/OrderServiceImpl.java]: Let’s divide this logic by some private methods for increasing readability

File: src/main/java/com/example/springbootbookshop/service/impl/OrderServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+    }
+
+    @Override
+    public OrderItemDto getOrderItem(Long userId, Long orderId, Long itemId) {
+        Set<Order> userOrders = orderRepository.findAllByUserId(userId);
+        Order userOrder = userOrders.stream()
+                .filter(order -> order.getId().equals(orderId))
+                .findFirst()
+                .orElseThrow(() -> new EntityNotFoundException(
+                        "User doen`t have order with id: " + orderId));
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/springbootbookshop/service/impl/OrderServiceImpl.java]: Why do you need this logic? Can you create a request to DB and find orderItem by input parameters?

File: src/main/java/com/example/springbootbookshop/service/impl/OrderServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+                new EntityNotFoundException("Item with id: " + itemId + " doesn`t found"));
+    }
+
+    @Override
+    public Set<OrderItemDto> getOrder(Long userId, Long orderId) {
+        Order orderItemsUser = orderRepository.findAllByUserId(userId).stream()
+                .filter(order -> order.getId().equals(orderId))
+                .findFirst()
+                .orElseThrow(() -> new EntityNotFoundException("No order with id: " + orderId
+                        + " for user with id: " + userId));
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/springbootbookshop/service/impl/OrderServiceImpl.java]: the same

File: src/main/java/com/example/springbootbookshop/service/impl/OrderServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+        return orderItemsRepository.findByOrder(orderItemsUser);
+    }
+
+    @Override
+    public Set<OrderDto> showAllOrders(Long userId) {
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/springbootbookshop/service/impl/OrderServiceImpl.java]:
```suggestion
    public Set<OrderDto> getAllOrders(Long userId) {
```

File: src/main/resources/db/changelog/changes/09-add-orders-table.yaml [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+                  defaultValueBoolean: false
+                  constraints:
+                    nullable: false
+        - addForeignKeyConstraint:
+            baseTableName: orders
+            baseColumnNames: user_id
+            referencedTableName: users
+            referencedColumnNames: id
+            constraintName: FK_orders_user
+            onDelete: CASCADE
```
Comment (liliia-ponomarenko) [inline src/main/resources/db/changelog/changes/09-add-orders-table.yaml]: add one empty line (check other files)

File: src/main/resources/db/changelog/changes/10-create-orders-items-table.yaml [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+            referencedTableName: orders
+            referencedColumnNames: id
+            constraintName: FK_order_items_order
+            onDelete: CASCADE
+        - addForeignKeyConstraint:
+            baseTableName: order_items
+            baseColumnNames: book_id
+            referencedTableName: books
+            referencedColumnNames: id
+            constraintName: FK_order_items_book
```
Comment (liliia-ponomarenko) [inline src/main/resources/db/changelog/changes/10-create-orders-items-table.yaml]: Do you have the same constraints for the entity? Should you have it?

File: src/main/resources/db/changelog/changes/09-add-orders-table.yaml [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+                  type: VARCHAR(255)
+                  constraints:
+                    nullable: false
+              - column:
+                  name: is_deleted
+                  type: BOOLEAN
+                  defaultValueBoolean: false
+                  constraints:
+                    nullable: false
+        - addForeignKeyConstraint:
```
Comment (liliia-ponomarenko) [inline src/main/resources/db/changelog/changes/09-add-orders-table.yaml]: the same

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Good job! Let’s improve your solution ;)

File: src/main/java/com/example/springbootbookshop/entity/OrderItem.java [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+@Data
+@Table(name = "order_items")
+@SQLDelete(sql = "UPDATE orders_items SET is_deleted = TRUE WHERE id = ?")
+@Where(clause = "is_deleted = false")
+public class OrderItem {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @ManyToOne(fetch = FetchType.LAZY)
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/springbootbookshop/entity/OrderItem.java]:
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
```

File: src/main/java/com/example/springbootbookshop/repository/OrderRepository.java [hunk 18]
```diff
@@ -0,0 +1,18 @@
...
+import com.example.springbootbookshop.entity.Order;
+import java.util.Optional;
+import java.util.Set;
+import org.springframework.data.jpa.repository.EntityGraph;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.stereotype.Repository;
+
+@Repository
+public interface OrderRepository extends JpaRepository<Order, Long> {
+
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/springbootbookshop/repository/OrderRepository.java]:
```suggestion
```

File: src/main/java/com/example/springbootbookshop/controller/OrderController.java [hunk 19]
```diff
@@ -0,0 +1,74 @@
...
+    @PreAuthorize("hasRole('ADMIN')")
+    public void updateStatus(@Positive @PathVariable Long orderId,
+                             @RequestBody UpdateRequestOrderStatus updateStatus) {
+        orderService.updateOrderStatus(orderId, updateStatus);
+    }
+
+    @GetMapping
+    @Operation(summary = "Get order history")
+    @PreAuthorize("hasRole('USER')")
+    public Set<OrderDto> getOrderHistory(Authentication authentication) {
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/springbootbookshop/controller/OrderController.java]: Pageable pageable ?

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Good job! Let's improve your solution!

**Review by liliia-ponomarenko — APPROVED** [review]
Great 😊

**Pull request merged**